thunderbird: add accountsOrder option

nfelber opened the pull request:
### Description

Fixes https://github.com/nix-community/home-manager/issues/5031 and adds `accountsOrder` option for declarative account ordering in Thunderbird's folder pane as discussed in the issue.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

@d-dervishi @jKarlson

alois31 commented on modules/programs/thunderbird.nix:
```diff
@@ -203,6 +203,23 @@ in {
+              accountsOrder = mkOption {
+                type = types.listOf types.str;
+                default = [ ];
+                description = ''
+                  Custom ordering of accounts in Thunderbird's folder pane.
+                  Enabled accounts that aren't listed here appear in an arbitrary
+                  order after the ordered accounts.
+                '';
```
This should explain how one has to specify the accounts for this option to actually work (by primary email address it seems). Adding this to the test would also be nice.

nfelber replied:
The accounts must be specified using their key  in `config.accounts.email.accounts`, as for `enabledAccounts`. I'll update the description to make it clearer.

alois31 commented on modules/programs/thunderbird.nix:
```diff
@@ -203,6 +203,23 @@ in {
+                  Enabled accounts that aren't listed here appear in an arbitrary
+                  order after the ordered accounts.
+                '';
+                example = ''
+                  [
+                    "account1"
+                    "account2"
+                    /* Other accounts in arbitrary order */
+                  ]
+                '';
```
This example is severely misleading since it suggests that the order is specified using the internal account ID. That said, I think it would be nice if something similar actually worked, since it would allow removing the special case for `account1` as well as enabling the user to use multiple imperative accounts.

nfelber replied:
I'm not sure what you mean here, although I agree that using the name `"account1"` is misleading. I'll change it. Can you elaborate a little?

alois31 replied:
That's what I meant exactly, since `account1` currently doesn't work (it would work in addition to the email address of configured accounts if you choose to implement the second solution below).

alois31 commented on modules/programs/thunderbird.nix:
```diff
@@ -415,14 +432,24 @@ in {
           a.thunderbird.profiles == [ ]
           || any (p: p == name) a.thunderbird.profiles) enabledAccountsWithId;
 
+        orderedAccountsIds = let
+          accountsOrderIds =
+            map (a: builtins.hashString "sha256" a) profile.accountsOrder;
+          enabledAccountsIds = map (a: a.id) accounts;
+        in (lists.intersectLists enabledAccountsIds accountsOrderIds)
+        ++ (lists.subtractLists accountsOrderIds enabledAccountsIds);
+
```
This silently discards accounts in the order that are missing. Such accounts should either throw an error, or (to enable the use of multiple imperative accounts as noted above) be included verbatim.

nfelber replied:
I agree and I can do either depending on what is better. I would tend towards the second option.

nfelber commented:
I implemented the discussed changes, let me know if it is better now. About the tests, I'd like to add one for the account order but I'm unsure why there are 3 accounts in `user_pref("mail.accountmanager.accounts", "...")` in the file `tests/modules/programs/thunderbird/thunderbird-expected-first-linux.js` while only 2 are declared in `tests/modules/programs/thunderbird/thunderbird.nix` and I don't want to break the current tests.

Also, I had a bit of an issue when trying to rebase my commit on to of upstream/main and made a bit of a mess (should be solved now), so I apologize for any trouble.

khaneliman commented on tests/modules/programs/thunderbird/thunderbird-expected-second-linux.js:
```diff
@@ -3,7 +3,7 @@
 user_pref("mail.account.account_bcd3ace52bed41febb6cdc2fb1303aebaa573e0d993872da503950901bb6c6fc.identities", "id_bcd3ace52bed41febb6cdc2fb1303aebaa573e0d993872da503950901bb6c6fc");
 user_pref("mail.account.account_bcd3ace52bed41febb6cdc2fb1303aebaa573e0d993872da503950901bb6c6fc.server", "server_bcd3ace52bed41febb6cdc2fb1303aebaa573e0d993872da503950901bb6c6fc");
-user_pref("mail.accountmanager.accounts", "account_bcd3ace52bed41febb6cdc2fb1303aebaa573e0d993872da503950901bb6c6fc");
+user_pref("mail.accountmanager.accounts", "account_bcd3ace52bed41febb6cdc2fb1303aebaa573e0d993872da503950901bb6c6fc,account1");
```
can you make these changes to the darwin test files, too

alois31 commented:
> I implemented the discussed changes, let me know if it is better now. About the tests, I'd like to add one for the account order but I'm unsure why there are 3 accounts in `user_pref("mail.accountmanager.accounts", "...")` in the file `tests/modules/programs/thunderbird/thunderbird-expected-first-linux.js` while only 2 are declared in `tests/modules/programs/thunderbird/thunderbird.nix` and I don't want to break the current tests.

The third account (c6cc42837ed0a8041f93ff12c579a4af0dbe702461c97eef069f9f5f8dc4bfab) is a feed account.

nfelber commented:
I added the `accountsOrder` option to the tests and mentioned feed accounts in the description of the option.

alois31 commented on modules/programs/thunderbird.nix:
```diff
@@ -277,6 +277,28 @@ in {
+                  in `config.accounts.email.accounts` (or
+                  `config.programs.thunderbird.profiles.<name>.feedAccounts` for feed
+                  accounts). Enabled accounts that aren't listed here appear in
+                  an arbitrary order after the ordered accounts.
```
… and local folders.

alois31 commented on modules/programs/thunderbird.nix:
```diff
@@ -573,13 +595,23 @@ in {
+          accountsOrderIds =
+            map (a: builtins.hashString "sha256" a) profile.accountsOrder;
+          enabledAccountsIds = map (a: a.id) accounts;
+        in accountsOrderIds
```
If the intent of this is to allow the ordering of imperative accounts, this is both undocumented and does not work as intended, as imperative accounts do not use the SHA256 hash in their name.

alois31 replied:
If this is fixed it would also be a good idea to not include the `account1` again below when it is already in the list.

nfelber commented:
Thanks for the review, good catches. I think I fixed the issues. I updated the tests to account for imperative accounts and the local folders as well. I tried to make the documentation not too confusing, but let me know if it is still not very clear.

khaneliman commented:
Sorry, could you rebase. Looks like I can't push to your branch.

nfelber commented:
Rebased
